### DISCUSSION TITLE
Handling of versioned jsonnet imports?

### DISCUSSION THREAD

**Discussion by strowi** [grafana/tanka#1228] [Q&A] [answered]
Hi all,

currently i have a few imports in use via jsonnet-bundler. These seem to be using subdirs instead of tags for versions.
For example cert-manager (similar for prometheus-operator etc..):

```json
{
  "version": 1,
  "dependencies": [
    {
      "source": {
        "git": {
          "remote": "https://github.com/jsonnet-libs/cert-manager-libsonnet.git",
          "subdir": "1.15"
        }
      },
      "version": "main"
    }
  ]
}
```

I'd be interested if there is a better way without using the fixed version in jsonnet-code to import those than doing this:

```
local cert_manager = import 'github.com/jsonnet-libs/cert-manager-libsonnet/1.15/main.libsonnet';
```

and having to replace all occurences.

How are you handling these?
And i also saw symlinks created in `vendor/$version` - what's the point of these? And what happens if i have 2 different vendors having the same version?

regards,
strowi

**Comment by zerok** [comment, marked as answer]
I doubt that there is a way to avoid hardcoding these somewhere since jsonnet does not support dynamic imports. What you can do, though, is collect all these into a single library and import the dependencies from there. This way you only have two places that need to be updated eventually instead of 1+n 🙂 

I think the best way to solve something like this would be through an extension of jsonnet-bundler to allow it to write to an arbitrary vendor folder. Haven't found a relevant ticket for that yet, though.

> And i also saw symlinks created in vendor/$version - what's the point of these? And what happens if i have 2 different vendors having the same version?

These are "legacy" import paths created by jsonnet-bundler. You can the creation of these symlinks by putting `"legacyImports": false` into your `jsonnetfile.json` 🙂

**Reply by strowi** [reply]
Thought so, already had a somewhat centra library for other stuff. I guess i will put it in there and have to do some renaming.

Must've missed the thing about legacy imports.. thx for the hint! ;)
